store/store.go

```diff
@@ -206,6 +208,11 @@ func (s *Store) Put(key []byte, value []byte) error {
 		// Two keys may point to same IndexKey (i.e. two CIDS same multihash),
 		// and they need to be treated as the same key.
 		if storedKey != nil {
+			// if we're not accepting updates, this is the point we bail --
+			// the identical key is in primary storage, we don't do update operations
+			if s.immutable {
+				return types.ErrKeyExists
+			}
```

Comment:
I think the index key is only the prefix. Now that you have retrieved `storedKey` (the full key), I think you need to compare `storedKey` to `key` to see if the full key stored under the `indexKey` is the same as `key` or is some other key that happens to have the same prefix.
```suggestion
			if s.immutable && bytes.Compare(key, storedKey) == 0 {
				return types.ErrKeyExists
			}
```

Reply:
as I read the code for `getPrimaryKeyData`, storedKey != nil indicates it has already done this check and verified equality. I will admit I find the code for `getPrimaryKeyData` a bit hard to parse.

Reply:
Yes, confirmed:
https://github.com/ipld/go-storethehash/blob/master/store/store.go#L335-L337